### PR TITLE
Approval button is missing after payment of the report with hold expense

### PR DESCRIPTION
**Pull request by ikevin127**
### Details

We're replacing the [logic](https://github.com/Expensify/App/blob/b732301c0011ae982744bf58c27c30fc3b62f6eb/src/libs/actions/IOU.ts#L6991-L6993) added by PR https://github.com/Expensify/App/pull/50817 by actually checking if the transaction has a receipt (scan) and is currently being scanned when it comes to hiding the `Approve` button because the previous logic was hiding the button regardless of the transaction being of type scan or not which shouldn't happen. The improved logic ensures we fix our issue while keeping the fix for the other issue for which the logic was written in the first place.

### Fixed Issues

$ https://github.com/Expensify/App/issues/50932
PROPOSAL: https://github.com/Expensify/App/issues/50932#issuecomment-2428153451

### Tests

Same as **QA Steps**.

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as **QA Steps**.

### QA Steps

**Prerequisite**: A workspace with an Employee and a general Approver (non-admin, can approve of any Employee).

1. Log in with your Approver account.
2. Log in with your Employee account on a different session (private tab or different browser).
3. As Employee: Go to your WS room and send two manual expenses.
4. As Employee: Hold one of the expenses.
5. As Approver: Approve just the pending amount (the expense which is not on hold).
6. As Approver: Verify that the held expense is now separate (below the approved one) and shows the `Approve` button.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

android

</details>

<details>
<summary>Android: mWeb Chrome</summary>

android-mweb

</details>

<details>
<summary>iOS: Native</summary>

ios

</details>

<details>
<summary>iOS: mWeb Safari</summary>

ios-mweb

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/8a8e1a6e-f35f-4f0f-9308-75bc9ca3d694

</details>

<details>
<summary>MacOS: Desktop</summary>

desktop

</details>

### CODE REVIEW

**Comment by jjcoffee** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

File: src/libs/actions/IOU.ts
Comment (jjcoffee) [inline src/libs/actions/IOU.ts:7040]: Ah there's a problem here, this still will allow approvals of 0-value scans (e.g. if you upload a junk receipt where the amount can't be read).

https://github.com/user-attachments/assets/10ea8cb7-91d5-4c19-90fe-36ed0c7ed473

Comment (ikevin127) [reply src/libs/actions/IOU.ts:7040]: @jjcoffee That's true and expected given my solution. As stated in the other issue which implemented the `unheldTotalIsZero` logic, the issue's (https://github.com/Expensify/App/issues/48590) the expected result states:

> Approver should not be able to approve scanning expense

None of the issues state that we shouldn't allow `Approve` when the transaction amount is 0, but that we shouldn't allow `Approve` while a scan transaction is currently scanning.

To confirm this I asked the C+ / CME in the other issue https://github.com/Expensify/App/issues/48590#issuecomment-2430420547 but didn't get a response so I assumed that the above expected result is the way to go.

**Why should we allow `Approve` when amount is 0 ?**
Because when a scan transaction with blurry image fails and amount is 0, you get an error on the field that you should add an amount, which allows you to add an amount if current is 0 then `Approve` - otherwise you deliberately `Approve` with 0 amount, on purpose.

##

If you have more context that I'm missing here which confirms that we shouldn't show the `Approve` button if the transaction amount is 0, please let me know and I'll add that to the PRs logic - otherwise, if there are no other issues with the logic I think we're good to move forward with this.

Comment (jjcoffee) [reply src/libs/actions/IOU.ts:7040]: > otherwise you deliberately Approve with 0 amount, on purpose.

@ikevin127 I can't think of any reason why you'd want to approve an expense with a 0 amount. There's nothing explicitly wrong with it but it just feels a bit messy that it's even possible, I think!

I'm also assuming we don't want to allow this both from the title of https://github.com/Expensify/App/issues/48590 as well as the PR's [test step](https://github.com/Expensify/App/pull/50016):

> verify that approver can't approve the scan expense as 0.00 amount

Comment (ikevin127) [reply src/libs/actions/IOU.ts:7040]: @jjcoffee The problem and pitfall in that PRs logic is that `canApproveIOU` is used in `MoneyReportHeader` and `ReportPreview` to show the `Approve` button for multiple expenses, because of this we can't hide `Approve` only for a transaction with amount 0 while showing it for the other transaction which is approvable (has all required fields completed).

That's why the `unheldTotalIsZero` logic added by the other PR was hiding the `Approve` button even for transactions what were not of type scan with 0 amount (our issue).

Say we implement this logic:

```ts
let isAmountMissing = false;
    let isTransactionBeingScanned = false;
    const reportTransactions = TransactionUtils.getAllReportTransactions(iouReport?.reportID);
    for (const transaction of reportTransactions) {
        const hasReceipt = TransactionUtils.hasReceipt(transaction);
        const isReceiptBeingScanned = TransactionUtils.isReceiptBeingScanned(transaction);

        // If transaction has receipt (scan) and its receipt is being scanned, we shouldn't be able to Approve
        if (hasReceipt && isReceiptBeingScanned) {
            isTransactionBeingScanned = true;
        }

        // Change isAmountMissing to true only if transaction amount is 0 (eg. failed scan)
        if (TransactionUtils.isAmountMissing(transaction)) {
            isAmountMissing = true;
        }
    }
    return isCurrentUserManager && !isOpenExpenseReport && !isApproved && !iouSettled && !isArchivedReport && !isTransactionBeingScanned && !isAmountMissing;
```
And we have a report with 2 transactions, 1 is manual and has amount and 1 is a failed scan (amount 0), with the logic above we won't see the `Approve` button at all.

Screenshot 2024-10-29 at 15 14 54

Whereas with the current logic of this PR, say the manual transaction is on hold and we also have 1 failed scan transaction (amount 0), we will see the `Approve` button on `ReportPreview` and `MoneyReportHeader`, which when clicked will ask the user if they want to approve amount 0 (only the failed scan transaction) or both (failed scan and the held manual $7), which nobody would do, meaning to approve only the $0, in a real-world scenario right ?

Screenshot 2024-10-29 at 15 19 26

To change this logic to your expectations would require rethinking how `canApproveIOU` works in `MoneyReportHeader` and `ReportPreview`, instead of showing `Approve` on a group of expenses (ReportPreview / MoneyReportHeader) we would have to implement logic to only show the button per individual transaction, which is definitely out of scope for our issue.

Please let me know what do you think we should do with this!

Comment (jjcoffee) [reply src/libs/actions/IOU.ts:7040]: Thanks for your thoughts @ikevin127! 

> we have a report with 2 transactions, 1 is manual and has amount and 1 is a failed scan (amount 0), with the logic above we won't see the Approve button at all.

Agree that we wouldn't want this, in this case and if the manual transaction is held I guess we'd ideally want to hide the `Approve only $0.00` button.

@marcochavezf Bringing you in here to chime in on whether we're okay to keep showing the `Approve` button for a failed scan (e.g. blurry image) with a **$0.00** amount. 

The main issue is that if you have another expense that's held, after tapping the `Approve` button, the first option shows as `Approve only $0.00`. That feels like dodgy UX to me, but I agree with @ikevin127 that fixing it is probably out of scope for this issue, since we wouldn't want to disable approving entirely as that would block approving the entire report (the second option in the screenshot below) and this would require a bit of a rethink of how to display that popup.

image

Just to note that the issue that introduced the regression had this as a test step:

> verify that approver can't approve the scan expense as 0.00 amount

So wanted to be sure we can go ahead despite that.

Comment (ikevin127) [reply src/libs/actions/IOU.ts:7040]: ## Breakdown - My two 🪙 

If you ask me, the changes of this PR are basically fixing our issue plus the expected result of this issue:
- https://github.com/Expensify/App/issues/48590

the problem is that the PR for the issue above:
- https://github.com/Expensify/App/pull/50817

went out of scope, because the issue stated clearly in its **expected result**:

> Approver should not be able to approve scanning expense

instead, the contributor tried to do more than the **expected result**, this is stated in PRs testing steps:

> verify that approver can't approve the scan expense as 0.00 amount

instead of simply fulfilling the issue's **expected result**. This led to our issue:
- https://github.com/Expensify/App/issues/50932

I'm assuming this happened because of lack of understanding of the codebase / logic and root cause.

##

**To conclude**: with our current logic it is not possible to only show the `Approve` button for individual transactions as I explained in https://github.com/Expensify/App/pull/51503#discussion_r1821606033, so this PR is both fixing its issue and the other issue where the **expected result** was to not be able to `Approve` a while an expense is still scanning.

From my proposal, this is how this would look:

| Can't approve while scanning |
| ------------- |
| <video src="https://github.com/user-attachments/assets/2da13c1c-1a73-45c3-a1aa-958313aea332"> |

##

If we want to rethink the logic and make the `Approve` button show per individual transaction, I think that deserves a separate task as it's way out of scope for any of the issues mentioned above.

cc @marcochavezf

**Comment by ikevin127** [comment]
cc @marcochavezf Bump for the discussion https://github.com/Expensify/App/pull/51503#discussion_r1818704296.

File: src/libs/actions/IOU.ts
Comment (marcochavezf) [reply src/libs/actions/IOU.ts:7040]: Thanks for bringing this up, guys! I agree that approving a report with an amount of 0 after scanning is not a good UX. Also, understandably it can be considered out-of-scope because it's an edge case we couldn't foresee. 

Given that we'd need to wait until this PR is deployed to staging/production to report it as a bug, what would be the solution to fix it here? We can adjust the price of the original issue by justifying the out-of-scope edge case.

Comment (ikevin127) [reply src/libs/actions/IOU.ts:7040]: @marcochavezf I appreciate the input 🙏 

> I agree that approving a report with an amount of 0 after scanning is not a good UX. Also, understandably it can be considered out-of-scope because it's an edge case we couldn't foresee.

1. Approving a report with an amount of 0 after scanning was allowed, present and possible on prod / staging even before PRs https://github.com/Expensify/App/pull/50817 changes and wasn't reported as an issue. Therefore the fact that the current PR fixes our issue while maintaining PRs https://github.com/Expensify/App/pull/50817 changes its a good thing which I could've left out as our issue's scope wasnt to improve or correct the other PRs logic but I did because otherwise that issue would've been reintroduced.
2. To reinforce point (1.), it is for sure out of scope since up until me bringing up the fact that this PRs fix handles both our issue and maintaining the other PRs logic, this wasn't even an issue (never reported as one).

**Given all this, here are the 2 roads I'm willing to move forward with right now regarding this PR:**
1. Merge this PR as is since it fixes both our issue while maintaining the expected reult fix for issue https://github.com/Expensify/App/issues/48590.
2. Drop everything, close the PR and leave it to somebody else to handle this + any out of scope work.

**Reasoning**
I'm not willing to go out of scope here regardless of extra compensation simply because the issue of `approving a report with an amount of 0 after scanning` was present on prod / staging even before PR https://github.com/Expensify/App/pull/50817, therefore it should not be a problem or reported as a regression of this PR - IMO it should be a `New Feature` issue since it requires lots of changes throughout the app.

Comment (jjcoffee) [reply src/libs/actions/IOU.ts:7040]: > Given that we'd need to wait until this PR is deployed to staging/production to report it as a bug, what would be the solution to fix it here?

Given that it's a bit of an edge case, and it was also previously possible (to approve a $0 amount), I think we can just proceed with this PR and report the bug after.

Comment (marcochavezf) [reply src/libs/actions/IOU.ts:7040]: Got it, thanks for the additional context @ikevin127! Yeah, if approving a report with an amount of 0 after scanning was already allowed then it makes sense to report it as a bug and continue with this PR as-is

**Review by marcochavezf — APPROVED** [review]
🚀

**Pull request merged**
